### PR TITLE
[WIP] Second checkpoint for alias analysis

### PR DESCRIPTION
**Pull request by mwu-tow**
### Pull Request Description
Code should be functionally complete with the exception of macro support.
PR against the previous accepted checkpoint for #322.

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Rust](https://github.com/luna/enso/blob/master/doc/rust-style-guide.md), [Scala](https://github.com/luna/enso/blob/master/doc/scala-style-guide.md), [Java](https://github.com/luna/enso/blob/master/doc/java-style-guide.md) or [Haskell](https://github.com/luna/enso/blob/master/doc/haskell-style-guide.md) style guides as appropriate.
- [x] All code has been tested where possible.

### CODE REVIEW

File: src/rust/ide/src/double_representation/alias_analysis.rs [hunk 1]
```diff
@@ -65,26 +72,257 @@ pub struct IdentifierUsage {
...
 
+
+// === Helper Datatypes ===
+
+/// Says whether the identifier occurrence introduces it into scope or uses it from scope.
+#[allow(missing_docs)]
+#[derive(Clone,Copy,Debug,Display,PartialEq)]
+pub enum OccurrenceKind { Used, Introduced }
+
+/// If the current context in the AST processor is a pattern context.
```
Comment (farmaazon) [inline src/rust/ide/src/double_representation/alias_analysis.rs:83]: I'd like to have a reference to a document which describes what pattern context is.

File: src/rust/ide/utils/src/vec.rs [hunk 2]
```diff
@@ -16,6 +16,17 @@ pub trait VecExt<T> : AsMut<Vec<T>> {
...
+    /// Removes the last `n` elements from the vector.
+    ///
+    /// Returns Some if elements were succesfully removed.
+    /// Returns None and leaves self intact, if there were not enough elements.
+    fn remove_last_n(&mut self, n:usize) -> Option<()> {
+        let vec      = self.as_mut();
+        let new_size = vec.len().checked_sub(n)?;
+        vec.truncate(new_size);
+        Some(())
+    }
```
Comment (farmaazon) [inline src/rust/ide/utils/src/vec.rs:29]: I find returning option for unit kinda, hmm... strange. Maybe return bool?
```suggestion
    /// Removes the last `n` elements from the vector.
    ///
    /// Returns Some if elements were succesfully removed.
    /// Returns None and leaves self intact, if there were not enough elements.
    fn remove_last_n(&mut self, n:usize) -> bool {
        {
            let vec      = self.as_mut();
            let new_size = vec.len().checked_sub(n)?;
            vec.truncate(new_size);
            Some(())
        }.is_some()
    }
```
Not sure if it actually compiles.

File: src/rust/Cargo.toml [hunk 3]
```diff
@@ -47,4 +47,4 @@ debug     = false
 [profile.test]
 opt-level = 0
 lto       = false
-debug     = false
+debug     = true
```
[the remark this answers is not in the document]
Comment (mwu-tow) [reply src/rust/Cargo.toml:50]: Ok, I need this to properly debug my test, it also provides line numbers in call stack when test is failing.

@wdanilo Can you comment why it was set to false? Is it still necessary?

File: src/rust/ide/src/double_representation/alias_analysis.rs (hunk 1 above)
Comment (mwu-tow) [reply src/rust/ide/src/double_representation/alias_analysis.rs:83]: Will add TODO comment, to be fixed when #349 lands.

**Pull request closed without merging**